### DISCUSSION TITLE
Building only GTK4 unstable

### DISCUSSION THREAD

**Discussion by DominusExult** [jralls/gtk-osx-build#111] [General]
I've tried building *only* GTK4 with `moduleset="https://gitlab.gnome.org/GNOME/gtk-osx/raw/master/modulesets-unstable/gtk-osx.modules"` and these commands:
```
jhbuild bootstrap-gtk-osx
jhbuild build meta-gtk-osx-bootstrap
jhbuild build gtk-4 
```
But this seems to be insufficient. The most glaring problem was that it didn't build the gdk-pixbuf loaders (except gif and tiff which it build). This causes of course to not display some theme elements and our program also makes use of xpm which also can't work without these.

Additionally, quick comparison shows these libs to be "missing" (some might not be needed, though)
```
libcroco-0.6.dylib
libgailutil-3.0.dylib
libgdk-3.0.dylib
librsvg-2.2.dylib
libxml2.2.dylib
```

Before bringing this up as an issue, I thought better to ask here in case I made an user error :)

**Comment by jralls** [comment]
The only loaders that gdk-pixbuf builds by default are png, jpeg, gif, and tiff, the first two of which are built-in. There's an optional "others" option that you can pass in your jhbuildrc-custom (`module_mesonargs['gdk-pixbuf'] = "-Dothers=enabled"`) to 'Enable other loaders, which are weakly maintained' as [meson_options.txt](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/meson_options.txt?ref_type=heads) says. If you only need a few you could add them as builtins with `-Dbuiltin_loaders=["jpeg", "png", "xpm"]` instead, see meson_options.txt for the list.
librsvg provides the svg loader.
libgdk-3.0 is obviously for Gtk3. librsvg is used by adwaita-icon-theme, which you may or may not need and librsvg is the only thing that depends on libxml2 any more. gettext builds its own libcroco now so we don't need it as a separate module. libgail was a Gtk2 accessibility library whose last release was in February 2011.

**Reply by DominusExult** [reply]
thanks, that helps me a lot.
More follow up questions:
- fundamental one: to update the gtk4 build (or have the mesonargs for gdk-pixbuf applied, is it enough to just run `jhbuild build gtk-4` again? (I was wondering about this before but since I often mucked around with my prefix and wanted to make sure I have a clean one I just nuked and rebuilt completely)
- building adwaita-icon-theme I'm instantly running into the problem that `git clone https://gitlab.gnome.org/GNOME/libxml2/2.11/libxml2-2.11.5.tar.xz` asks for gitlab.gnome.org credentials and fails because I don't have permissions to view it.
```
Loading .env environment variables...
*** Checking out libxml2 *** [1/25]
git clone https://gitlab.gnome.org/GNOME/libxml2/2.11/libxml2-2.11.5.tar.xz
Cloning into 'libxml2-2.11.5.tar.xz'...
remote: The project you were looking for could not be found or you don't have permission to view it.
fatal: repository 'https://gitlab.gnome.org/GNOME/libxml2/2.11/libxml2-2.11.5.tar.xz.git/' not found
```

**Reply by jralls** [reply]
For the first, it depends, see https://gnome.pages.gitlab.gnome.org/jhbuild/config-reference.html#build-policy. You can also force a rebuild with `-f`. Since you only need to rebuild gdk-pixbuf you can run `jhbuild buildone -f -c gdk-pixbuf`.

For the second, oops. git can't clone a tarball. I just pushed a fix.

**Reply by DominusExult** [reply]
:)
Another problem in configure phase of librsvg:
```
*** Configuring librsvg *** [24/25]
meson --prefix /opt/gtk4 --libdir lib -Dintrospection=disabled -Ddocs=disabled --wrap-mode=nofallback /Users/Dominus/gtk/source/librsvg
The Meson build system
Version: 1.4.1
Source dir: /Users/Dominus/gtk/source/librsvg
Build dir: /Users/Dominus/.cache/jhbuild/build/librsvg
Build type: native build
Project name: librsvg
Project version: 2.58.91
C compiler for the host machine: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang (clang 15.0.0 "Apple clang version 15.0.0 (clang-1500.3.9.4)")
C linker for the host machine: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang ld64 1053.12
Host machine cpu family: aarch64
Host machine cpu: aarch64
Program cargo found: YES 1.79.0 1.79.0 (/Users/Dominus/.new_local/bin/cargo)
Program meson/cargo_wrapper.py found: YES (/Users/Dominus/gtk/source/librsvg/meson/cargo_wrapper.py)
Program cargo-cbuild found: NO

../../../../gtk/source/librsvg/meson.build:29:10: ERROR: Program 'cargo-cbuild' not found or not executable

A full log can be found at /Users/Dominus/.cache/jhbuild/build/librsvg/meson-logs/meson-log.txt
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.
*** Error during phase configure of librsvg: ########## Error running meson --prefix /opt/gtk4 --libdir lib -Dintrospection=disabled -Ddocs=disabled --wrap-mode=nofallback /Users/Dominus/gtk/source/librsvg *** [24/25]
```

**Reply by jralls** [reply]
Oh, right, I forgot about that one. I haven't figured out how to automate it. From a jhbuild shell run
```
cargo install cargo-c --features=vendored-openssl
```
then quit the shell and re-run configuration.

**Reply by DominusExult** [reply]
Thank you, getting there :)

Your libxml2 fix was incorrect, though. it should be `repo="ftp.gnome.org"` AFAICT (jusdging from other `repo=` cases and this worked right now for me)

**Reply by DominusExult** [reply]
So, updating gdk-pixbuf a success, and installing adwaita-icon-theme a success as well.

But the svg lib has another little problem: it adds `libpixbufloader_svg.dylib` to gdk-pixbuf loaders instead of an `.so` file (which will later confuse gtk-mac-bundler as it will only copy `.so` files by default).
*AND* AFAICT it doesn't update the gdk-pixbuf loaders cache with the svg stuff (which can be done manually and is mentioned in the build logs, if you know what to look for).

**Reply by jralls** [reply]
> Your libxml2 fix was incorrect, though.

Well, half-wrong (modulesets was right), and not the way you thought: It's download.gnome.org in every other moduleset so I changed that one to match. Thanks for telling me.

**Comment by DominusExult** [comment]
Desperately :) trying to rebuild GTK-4 unstable I've run into a problem with  `glib-no-introspection`:

` ../../../../gtk/source/glib/meson.build:960:2: ERROR: Value "gnu99,c99" (of type "string") for combo option "C language standard to use" is not one of the choices. Possible choices are (as string): "none", "c89", "c99", "c11", "c17", "gnu89", "gnu99", "gnu11", "gnu17".`

As above my only changes to jhbuildrc-custom are:
```
module_mesonargs['gdk-pixbuf'] = "-Dothers=enabled"
prefix = '/opt/gtk4'
setup_sdk(target="14.5")
moduleset="https://gitlab.gnome.org/GNOME/gtk-osx/raw/master/modulesets-unstable/gtk-osx.modules"
```
Any idea what is happening here? (my guess is that master of glib has an issue)

**Reply by DominusExult** [reply]
setting glib to version 2.82.0 instead of main, makes it work. 

However I ran into another problem, for some reason the symlink bin/pkgconf -> bin/pkg-config wasn't working (the pkg-config symlink didn't point to anything)and thus glib configure wouldn't find libpcre2. This was with a new gtk4 prefix, no idea why that failed. Fixing that made everything run through.

**Reply by jralls** [reply]
https://gitlab.gnome.org/GNOME/glib/-/commit/51e3e7d9ae330cd798604d0bd1ac70d0daa93681 that was merged last Thursday changes meson.build line 8 from `c_std=gnu99` to `c_std=gnu99,c99` which is the source of the error. But it also changes the meson version requirement to 1.40.0 and according to https://mesonbuild.com/Builtin-options.html that's legal since meson 1.30.
The error message indicates that your meson is older than 1.30, but then why didn't meson.build error out because the version is too low?

**Reply by DominusExult** [reply]
running `meson -v` in `jhbuild shell` returns 1.4.1 but maybe there are some dependencies that were not updated and required for this.

Edit: the only step I skipped was running "gtk-osx-setup.sh" as I already had jhbuild setup before, so maybe there is something too old, even though meson was up to date.

**Reply by jralls** [reply]
>  for some reason the symlink bin/pkgconf -> bin/pkg-config wasn't working

Probably because it's supposed to go the other way. It should be created correctly by pkgconf with https://gitlab.gnome.org/GNOME/gtk-osx/-/blob/master/patches/pkgconf-link-pkg-config.patch?ref_type=heads applied.

What do `which meson` and `which python3` run in jhbuild shell return?

**Reply by DominusExult** [reply]
> > for some reason the symlink bin/pkgconf -> bin/pkg-config wasn't working
> 
> Probably because it's supposed to go the other way. It should be created correctly by pkgconf with https://gitlab.gnome.org/GNOME/gtk-osx/-/blob/master/patches/pkgconf-link-pkg-config.patch?ref_type=heads applied.

I might have written it the wrong way around. pkgconf is the binary and pkg-config should be the symlink, right?. This wasn't working for some reason.

> What do `which meson` and `which python3` run in jhbuild shell return?

/Users/Dominus/.new_local/share/virtualenvs/etc-Mw2UFQW9/bin/meson

/Users/Dominus/.new_local/share/virtualenvs/etc-Mw2UFQW9/bin/python3
(version 3.11.7, btw)

**Reply by jralls** [reply]
The patch uses meson to create the symlink and it comes out perhaps a bit goofy:
```
pkg-config@ -> /Users/john/Development/gtk-unstable-11-arm64/build/pkgconf-2.2.0/pkgconf
```
So if you were to clean up your build directory it would fail.

**Reply by DominusExult** [reply]
Ah, yes, that explains it (especially as I found the same problem in my GTK3 prefix). It wasn't a problem after I build the prefixes since I use two prefixes together, the GTK3 one and another for non-GTK-ON-OSX libs and I pointed to that pkg-config.

**Comment by DominusExult** [comment]
When you create a clean Gtk4 prefix I found that it doesn't install the hicolor theme.
As you pointed out to me in https://gitlab.gnome.org/GNOME/gtk-osx/-/issues/94 Gtk4 comes with many icons but gtk-mac-bundler really wants to install the hicolor theme and errors out when the theme file is not present.

```
Traceback (most recent call last):
  File "/opt/arm64/bin/gtk-mac-bundler", line 6, in <module>
    bundler.main.main(sys.argv[1:])
  File "/Users/dominus/code/git/gtk-mac-bundler/bundler/main.py", line 19, in main
    bundler.run()
  File "/Users/dominus/code/git/gtk-mac-bundler/bundler/bundler.py", line 424, in run
    self.copy_icon_themes()
  File "/Users/dominus/code/git/gtk-mac-bundler/bundler/bundler.py", line 335, in copy_icon_themes
    theme.copy_target(self.project)
  File "/Users/dominus/code/git/gtk-mac-bundler/bundler/project.py", line 427, in copy_target
    super().copy_target(the_project)
  File "/Users/dominus/code/git/gtk-mac-bundler/bundler/project.py", line 186, in copy_target
    source = self.compute_source_path(the_project)
  File "/Users/dominus/code/git/gtk-mac-bundler/bundler/project.py", line 179, in compute_source_path
    raise ValueError("Cannot find source to copy: " + source)
ValueError: Cannot find source to copy: /opt/gtk4/share/icons/hicolor/index.theme
```

Adding the theme via `jhbuild build hicolor-icon-theme` fixes that but I guess either the bundler script needs updating (if/since theme is not needed) or the hicolor theme needs to be a dependency of the module gtk-4.

And related, wouldn't it make sense to have the cache files created when you install the themes modules (hiclor, adwaita)? Sure, running `gtk4-update-icon-cache path_to_theme` isn't that much of a hassle :)

**Reply by jralls** [reply]
That's actually a relic in gtk-mac-bundler. Gtk2 and Gtk3 required the hicolor theme so installing it is hard-coded in gtk-mac-bundler's project.py. I need to figure out a way to make it optional without breaking everyone's Gtk3 bundle files.

I've filed [an issue](https://gitlab.gnome.org/GNOME/gtk-mac-bundler/-/issues/27) and [another](https://gitlab.gnome.org/GNOME/gtk-mac-bundler/-/issues/28) for the icon cache.

**Reply by DominusExult** [reply]
I guess since the bundle.py already has the ability to differ by `self.meta.gtk == 'gtk4'` it might not be that much of a hassle (not counting the time that someone needs to spend on it).

Seems it is my yearly time of messing with my working prefixes and running into issues (or non issues :))

**Comment by DominusExult** [comment]
@jralls can you shed some more light on GTK4, svg and icon sets?
At https://gitlab.gnome.org/GNOME/gtk-osx/-/issues/94#note_2416045 you wrote:

> gtk-4 has a ton of svg icons but it uses a [cairo svgsurface](https://www.cairographics.org/manual/cairo-SVG-Surfaces.html) to render them; librsvg isn't required.

So, I can see the svg icons in GTK4's source but where do these end up in the prefix?

We are using document-edit and document-new and thus I found out that for these to be not just icon placeholders, we need an icon set that has these (e.g. Adwaita). But everything I tried pointed me to be needing to build librsvg to use icon sets other than the built-in one.

**Reply by jralls** [reply]
@DominusExult it appears to me that only the few icons associated with the demo programs are installed. They go into $PREFIX/share/icons/hicolor. The rest seem to be ignored. Using gtk4-demo I looked at some of the demo sub-programs and found that there were lots of missing icons; installing adwaita-icon-theme provided some but not all so that's not really the answer.

Yes, adwaita-icon-theme requires librsvg, so if that does the job for you by all means add it to your module list.

**Comment by DominusExult** [comment]
gdk-pixbuf in unstable seems to depend on glycin now.

```
Run-time dependency glycin-1 found: NO (tried pkgconfig, framework and cmake)

../../../../gtk/source/gdk-pixbuf/meson.build:284:15: ERROR: Dependency "glycin-1" not found, tried pkgconfig, framework and cmake
```

**Reply by jralls** [reply]
Fortunately it's an option; unfortunately it has to be explicitly disabled. [Glycin's README](https://gitlab.gnome.org/GNOME/glycin/-/blob/main/README.md?ref_type=heads) says that it's Linux only.

**Reply by jralls** [reply]
I've pushed a change to `modulesets-unstable/gtk-osx.modules` that disables glycin for gdk-pixbuf.

**Reply by DominusExult** [reply]
thanks.

**Reply by DominusExult** [reply]
There was one other issue (actually before it reached the glycin issue). Building fontconfig, it somehow couldn't find 'fcstdint.h'.
```
*** Building fontconfig *** [16/31]
ninja  
[7/115] Compiling C object test/test_ptrlist.p/test-ptrlist.c.o
FAILED: test/test_ptrlist.p/test-ptrlist.c.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -Itest/test_ptrlist.p -Itest -I../../../../gtk/source/fontconfig/test -I. -I../../../../gtk/source/fontconfig -Isrc -I../../../../gtk/source/fontconfig/src -I/opt/gtk4/include -fdiagnostics-color=always -Wall -Winvalid-pch -std=c11 -O2 -g -DHAVE_CONFIG_H -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=11.1 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -MD -MQ test/test_ptrlist.p/test-ptrlist.c.o -MF test/test_ptrlist.p/test-ptrlist.c.o.d -o test/test_ptrlist.p/test-ptrlist.c.o -c ../../../../gtk/source/fontconfig/test/test-ptrlist.c
In file included from ../../../../gtk/source/fontconfig/test/test-ptrlist.c:5:
../../../../gtk/source/fontconfig/src/fcint.h:32:10: fatal error: 'fcstdint.h' file not found
   32 | #include "fcstdint.h"
      |          ^~~~~~~~~~~~
1 error generated.
[24/115] Generating fc-lang/fclang.h with a custom command
ninja: build stopped: subcommand failed.
```
After copying this and the other `fc*.h` files from `$HOME/.cache/jhbuild/build/fontconfig` to `$HOME/gtk/source/fontconfig-2.16.0` I was able to continue and finish building.

**Reply by jralls** [reply]
No need to copy it, it's already created. You can just re-run the build and it will work, there seems to be a race condition between copying it from the srcdir's `fcstdint.h.in` and building test-ptrlist.

**Reply by DominusExult** [reply]
🤦‍♂️I was wondering why it worked copying these to the sources
